tests/ppc_disasm.py

```diff
@@ -0,0 +1,568 @@
+# and labels in the output asm only use alphanumerics and '$'
+SUBSTITUTIONS: Tuple[Tuple[str, str], ...] = (
+    ("<", "$$0"),
+    (">", "$$1"),
+    ("@", "$$2"),
+    ("\\", "$$3"),
+    (",", "$$4"),
+    ("-", "$$5"),
+)
+
+
+def symbol_name(sym: ElfSymbol) -> str:
+    name = sym.name
+    for sub in SUBSTITUTIONS:
+        name = name.replace(*sub)
+    return name
+
+
+def address_label(addr: int) -> str:
+    return f".L{addr:08X}"
+
+
+# Returns true if the instruction is a load or store with the given register as a base
+def is_load_store_reg_offset(insn: CsInsn, reg: Optional[CsRegister]) -> bool:
+    return insn.id in LOAD_STORE_MNEMONICS and (
+        reg == None or insn.operands[1].mem.base == reg
+    )
+
+
+# Converts the instruction to a string, fixing various issues with Capstone
+def instruction_to_text(insn: CsInsn, raw: int, section: ElfSection) -> Optional[str]:
+    # Look up the relocation for this address, which is either at offset+0 or offset+2
+    # (It should be possible to determine which offset to use from either the reolcation
+    # type or the instruction mnemonic, but for now just check both since they should
+    # be mutually exclusive.)
+    offset = insn.address - section.address
+    reloc = section.relocations.get(offset)
+    if reloc is None:
+        reloc = section.relocations.get(offset + 2)
+
+    label: Optional[str] = None
+    if reloc is not None:
+        label = symbol_name(reloc.symbol)
+    elif insn.operands:
+        label_addr = insn.operands[0].imm
+        symbol = section.symbols.get(label_addr)
+        if symbol is not None:
+            label = symbol_name(symbol)
+        else:
+            label = address_label(label_addr)
+
+    # Probably data, not a real instruction
+    if insn.id == cs.ppc.PPC_INS_BDNZ and (insn.bytes[0] & 1):
+        return None
+    if insn.id in {
+        cs.ppc.PPC_INS_B,
+        cs.ppc.PPC_INS_BL,
+        cs.ppc.PPC_INS_BDZ,
+        cs.ppc.PPC_INS_BDNZ,
+    }:
+        if not label:
+            return "%s %s" % (
+                insn.mnemonic,
+                f"unk label: {(raw, insn.address - section.address)}",
+            )
+            return None
+        return "%s %s" % (insn.mnemonic, label)
+    elif insn.id == cs.ppc.PPC_INS_BC:
+        branchPred = "+" if (insn.bytes[1] & 0x20) else ""
+        if not insn.operands:
+            pass
+        elif insn.operands[0].type == cs.ppc.PPC_OP_IMM:
+            if not label:
+                return None
+            return "%s%s %s" % (
+                insn.mnemonic,
+                branchPred,
+                label,
+            )
+        elif insn.operands[1].type == cs.ppc.PPC_OP_IMM:
+            if not label:
+                return None
+            return "%s%s %s, %s" % (
+                insn.mnemonic,
+                branchPred,
+                insn.reg_name(insn.operands[0].value.reg),
+                label,
+            )
+    if reloc is not None:
```

Comment:
the code here looks a bit backward to me: it makes inferences on reloc kind based on circumstantial evidence and assumes sda21/l/h/ha are the only kinds that can occur. Would be nicer to have a comprehensive switch over reloc.relocation_type and error out if it fails to match anything.

(totally fine if you don't want to spend time cleaning this up though seeing as it's (at least for now) test code, and if we turn it into non-test code we'll want to carefully review/rewrite it anyway)

Reply:
Yeah - the original doldisasm.py *had* to make these inferences, because it was working from the linked output. 
Our job here is much easier because we have the reloc types. 

I originally tried to get away with minimal changes to the code in this region, but at this point it's barely recognizable :smile: 
I agree having a switch over `relocation_type` would be much much better - but I think I'll punt on that for now.